### PR TITLE
Close SocketChannel upon exceptions to avoid leaking

### PR DESCRIPTION
**Pull request by blindpirate**
This fixes https://github.com/netty/netty/issues/10109
by correctly closing the channel upon exceptions.

Fixes https://github.com/netty/netty/issues/10109

### CODE REVIEW

**Comment by normanmaurer** [comment]
@blindpirate please sign our ICLA and let me know once done:

https://netty.io/s/icla

**Comment by blindpirate** [comment]
Done. @normanmaurer

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: transport/src/main/java/io/netty/bootstrap/Bootstrap.java [hunk 1]
```diff
@@ -189,7 +189,13 @@ private ChannelFuture doResolveAndConnect0(final Channel channel, SocketAddress
                                                final SocketAddress localAddress, final ChannelPromise promise) {
         try {
             final EventLoop eventLoop = channel.eventLoop();
-            final AddressResolver<SocketAddress> resolver = this.resolver.getResolver(eventLoop);
+            AddressResolver<SocketAddress> resolver;
+            try {
+                resolver = this.resolver.getResolver(eventLoop);
+            } catch (Throwable e) {
+                channel.close();
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/bootstrap/Bootstrap.java]: @blindpirate I think this should be changed to:

```
try {
    ...
} catch (Throwable cause) {
    channel.close();
    promise.setFailure(e);
    return;
}

```

Comment (normanmaurer) [reply transport/src/main/java/io/netty/bootstrap/Bootstrap.java]: Also would be nice if you could write a unit test for it and include it in `BootstrapTest`.

Comment (blindpirate) [reply transport/src/main/java/io/netty/bootstrap/Bootstrap.java]: OK, will try.

Comment (normanmaurer) [reply transport/src/main/java/io/netty/bootstrap/Bootstrap.java]: @blindpirate please ping me once done ...

Comment (normanmaurer) [reply transport/src/main/java/io/netty/bootstrap/Bootstrap.java]: actually let me just take care and add me as an co-author then.

**Comment by normanmaurer** [comment]
@blindpirate I just pushed a commit with the changes and a test case... PTAL

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by blindpirate** [comment]
@normanmaurer out of curiosity, is "test this please" is a fixed command for @netty-bot ? Can I ask him to test by "can you please test this change"? Is that bot opensource?

**Comment by blindpirate** [comment]
@normanmaurer out of curiosity, is "test this please" is a fixed command for @netty-bot ? Can I ask him to test by "can you please test this change"? Is that bot opensource?

**Comment by normanmaurer** [comment]
@blindpirate it can only be done by admins so we ensure no-one DDOS it. It is a Jenkins plugin.

**Comment by blindpirate** [comment]
Good to know, thanks!

**Pull request merged**